File: dot-config/swaync/scripts/domain/ports/verify.py
"""Verify an adapter satisfies a domain port contract.

Fail loud: the first adapter that drifts from its port aborts with a precise
message, mirroring ``hypr/domain/ports/verify.lua``. The port protocols are
re-exported here so the composition root can name them.
"""

from __future__ import annotations

from typing import Protocol

from domain.ports.artifacts import ConfigWriter, StyleWriter
from domain.ports.core import CommandRunner, LifetimePort, Logger, TimePort

__all__ = [
    "CommandRunner",
    "ConfigWriter",
    "LifetimePort",
    "Logger",
    "StyleWriter",
    "TimePort",
    "assert_port",
    "assert_ports",
]


def assert_port(adapter: object, port: type[Protocol]) -> None:
    if not isinstance(adapter, port):
        raise TypeError(f"{type(adapter).__name__} does not implement {port.__name__}")


def assert_ports(adapter: object, ports: tuple[type[Protocol], ...]) -> None:
    for port in ports:
        assert_port(adapter, port)

File: dot-config/hypr/domain/ports/verify.lua
-- Verify an adapter satisfies a domain port contract.
-- Fail loud: the first missing method aborts with a precise message.

local M = {}

M.monitor = require("domain.ports.monitor")
M.environment = require("domain.ports.environment")
M.input = require("domain.ports.input")
M.binding = require("domain.ports.binding")
M.command = require("domain.ports.command")
M.visual = require("domain.ports.visual")
M.layer = require("domain.ports.layer")
M.window_rule = require("domain.ports.window_rule")
M.runtime = require("domain.ports.runtime")
M.wallpaper = require("domain.ports.wallpaper")
M.logger = require("domain.ports.logger")
M.time = require("domain.ports.time")
M.lifetime = require("domain.ports.lifetime")

-- Ports implemented by the single Hyprland adapter (one external system).
M.HYPR_PORTS = {
  M.monitor,
  M.environment,
  M.input,
  M.binding,
  M.command,
  M.visual,
  M.layer,
  M.window_rule,
  M.runtime,
}

function M.assert_port(adapter, port)
  for name, value in pairs(port) do
    if type(value) == "function" then
      assert(
        type(adapter[name]) == "function",
        string.format("Adapter does not implement %s:%s", port.NAME, name)
      )
    end
  end
end

function M.assert_ports(adapter, ports)
  for _, port in ipairs(ports) do
    M.assert_port(adapter, port)
  end
end

return M
